File: packages/sdk/go/devbox/environments.go
package devbox

import (
	"context"
	"fmt"
	"net/url"
	"time"
)

// EnvironmentsService handles environment operations
type EnvironmentsService struct {
	client *Client
}

// Environment represents a development environment
type Environment struct {
	ID             string            `json:"id"`
	Name           string            `json:"name"`
	Description    string            `json:"description,omitempty"`
	TemplateID     string            `json:"template_id"`
	Status         string            `json:"status"`
	Resources      *ResourceConfig   `json:"resources,omitempty"`
	EnvVars        map[string]string `json:"env_vars,omitempty"`
	Ports          []PortMapping     `json:"ports,omitempty"`
	PreviewURL     string            `json:"preview_url,omitempty"`
	SSHHost        string            `json:"ssh_host,omitempty"`
	SSHPort        int               `json:"ssh_port,omitempty"`
	CreatedAt      time.Time         `json:"created_at"`
	UpdatedAt      time.Time         `json:"updated_at"`
	LastActiveAt   *time.Time        `json:"last_active_at,omitempty"`
}

// ResourceConfig defines resource allocation
type ResourceConfig struct {
	CPU     string `json:"cpu"`
	Memory  string `json:"memory"`
	Storage string `json:"storage"`
}

// PortMapping defines a port mapping
type PortMapping struct {
	ContainerPort int    `json:"container_port"`
	Protocol      string `json:"protocol"`
	Public        bool   `json:"public"`
}

// SSHConfig contains SSH connection information
type SSHConfig struct {
	Host          string `json:"host"`
	Port          int    `json:"port"`
	Username      string `json:"username"`
	PrivateKey    string `json:"private_key"`
	ConfigSnippet string `json:"config_snippet"`
}

// EnvironmentMetrics contains environment metrics
type EnvironmentMetrics struct {
	CPUUsage     []MetricPoint `json:"cpu_usage"`
	MemoryUsage  []MetricPoint `json:"memory_usage"`
	StorageUsage []MetricPoint `json:"storage_usage"`
	NetworkRx    []MetricPoint `json:"network_rx"`
	NetworkTx    []MetricPoint `json:"network_tx"`
}

// MetricPoint represents a single metric data point
type MetricPoint struct {
	Timestamp time.Time `json:"timestamp"`
	Value     float64   `json:"value"`
}

// CreateEnvironmentRequest contains parameters for creating an environment
type CreateEnvironmentRequest struct {
	Name        string            `json:"name"`
	Description string            `json:"description,omitempty"`
	TemplateID  string            `json:"template_id"`
	Resources   *ResourceConfig   `json:"resources,omitempty"`
	EnvVars     map[string]string `json:"env_vars,omitempty"`
	Ports       []PortMapping     `json:"ports,omitempty"`
	GitRepoURL  string            `json:"git_repo_url,omitempty"`
}

// UpdateEnvironmentRequest contains parameters for updating an environment
type UpdateEnvironmentRequest struct {
	Name        string            `json:"name,omitempty"`
	Description string            `json:"description,omitempty"`
	Resources   *ResourceConfig   `json:"resources,omitempty"`
	EnvVars     map[string]string `json:"env_vars,omitempty"`
}

// EnvironmentListOptions contains options for listing environments
type EnvironmentListOptions struct {
	ListOptions
	Status     string
	TemplateID string
}

// EnvironmentListResponse contains a paginated list of environments
type EnvironmentListResponse struct {
	Data       []Environment `json:"data"`
	Pagination Pagination    `json:"pagination"`
}

// List returns a list of environments
func (s *EnvironmentsService) List(ctx context.Context, opts *EnvironmentListOptions) (*EnvironmentListResponse, error) {
	path := "/environments"
	if opts != nil {
		params := url.Values{}
		if opts.Page > 0 {
			params.Set("page", fmt.Sprintf("%d", opts.Page))
		}
		if opts.PageSize > 0 {
			params.Set("page_size", fmt.Sprintf("%d", opts.PageSize))
		}
		if opts.Status != "" {
			params.Set("status", opts.Status)
		}
		if opts.TemplateID != "" {
			params.Set("template_id", opts.TemplateID)
		}
		if len(params) > 0 {
			path += "?" + params.Encode()
		}
	}

	var resp EnvironmentListResponse
	if err := s.client.request(ctx, "GET", path, nil, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// Get returns an environment by ID
func (s *EnvironmentsService) Get(ctx context.Context, id string) (*Environment, error) {
	var env Environment
	if err := s.client.request(ctx, "GET", "/environments/"+id, nil, &env); err != nil {
		return nil, err
	}
	return &env, nil
}

// Create creates a new environment
func (s *EnvironmentsService) Create(ctx context.Context, req *CreateEnvironmentRequest) (*Environment, error) {
	var env Environment
	if err := s.client.request(ctx, "POST", "/environments", req, &env); err != nil {
		return nil, err
	}
	return &env, nil
}

// Update updates an environment
func (s *EnvironmentsService) Update(ctx context.Context, id string, req *UpdateEnvironmentRequest) (*Environment, error) {
	var env Environment
	if err := s.client.request(ctx, "PUT", "/environments/"+id, req, &env); err != nil {
		return nil, err
	}
	return &env, nil
}

// Delete deletes an environment
func (s *EnvironmentsService) Delete(ctx context.Context, id string) error {
	return s.client.request(ctx, "DELETE", "/environments/"+id, nil, nil)
}

// Start starts an environment
func (s *EnvironmentsService) Start(ctx context.Context, id string) (*Environment, error) {
	var env Environment
	if err := s.client.request(ctx, "POST", "/environments/"+id+"/start", nil, &env); err != nil {
		return nil, err
	}
	return &env, nil
}

// Stop stops an environment
func (s *EnvironmentsService) Stop(ctx context.Context, id string) (*Environment, error) {
	var env Environment
	if err := s.client.request(ctx, "POST", "/environments/"+id+"/stop", nil, &env); err != nil {
		return nil, err
	}
	return &env, nil
}

// Restart restarts an environment
func (s *EnvironmentsService) Restart(ctx context.Context, id string) (*Environment, error) {
	var env Environment
	if err := s.client.request(ctx, "POST", "/environments/"+id+"/restart", nil, &env); err != nil {
		return nil, err
	}
	return &env, nil
}

// GetSSHConfig returns SSH configuration for an environment
func (s *EnvironmentsService) GetSSHConfig(ctx context.Context, id string) (*SSHConfig, error) {
	var config SSHConfig
	if err := s.client.request(ctx, "GET", "/environments/"+id+"/ssh-config", nil, &config); err != nil {
		return nil, err
	}
	return &config, nil
}

// GetMetrics returns metrics for an environment
func (s *EnvironmentsService) GetMetrics(ctx context.Context, id string, startTime, endTime *time.Time) (*EnvironmentMetrics, error) {
	path := "/environments/" + id + "/metrics"
	params := url.Values{}
	if startTime != nil {
		params.Set("start_time", startTime.Format(time.RFC3339))
	}
	if endTime != nil {
		params.Set("end_time", endTime.Format(time.RFC3339))
	}
	if len(params) > 0 {
		path += "?" + params.Encode()
	}

	var metrics EnvironmentMetrics
	if err := s.client.request(ctx, "GET", path, nil, &metrics); err != nil {
		return nil, err
	}
	return &metrics, nil
}

File: packages/sdk/python/clouddevbox/environments.py
"""Environments Service"""

from dataclasses import dataclass
from typing import Optional, TYPE_CHECKING
from datetime import datetime

from .types import ResourceConfig, PortMapping

if TYPE_CHECKING:
    from .client import DevBoxClient, AsyncDevBoxClient


@dataclass
class Environment:
    """Development environment"""
    id: str
    name: str
    template_id: str
    status: str
    created_at: str
    updated_at: str
    description: Optional[str] = None
    resources: Optional[ResourceConfig] = None
    env_vars: Optional[dict[str, str]] = None
    ports: Optional[list[PortMapping]] = None
    preview_url: Optional[str] = None
    ssh_host: Optional[str] = None
    ssh_port: Optional[int] = None
    last_active_at: Optional[str] = None


@dataclass
class SSHConfig:
    """SSH connection configuration"""
    host: str
    port: int
    username: str
    private_key: str
    config_snippet: str


@dataclass
class MetricPoint:
    """Single metric data point"""
    timestamp: str
    value: float


@dataclass
class EnvironmentMetrics:
    """Environment metrics"""
    cpu_usage: list[MetricPoint]
    memory_usage: list[MetricPoint]
    storage_usage: list[MetricPoint]
    network_rx: list[MetricPoint]
    network_tx: list[MetricPoint]


class EnvironmentsService:
    """Environments service for sync client"""

    def __init__(self, client: "DevBoxClient"):
        self._client = client

    def list(
        self,
        page: int = 1,
        page_size: int = 20,
        status: Optional[str] = None,
        template_id: Optional[str] = None,
    ) -> dict:
        """List environments"""
        params = {"page": page, "page_size": page_size}
        if status:
            params["status"] = status
        if template_id:
            params["template_id"] = template_id
        return self._client.request("GET", "/environments", params=params)

    def get(self, id: str) -> dict:
        """Get environment by ID"""
        return self._client.request("GET", f"/environments/{id}")

    def create(
        self,
        name: str,
        template_id: str,
        description: Optional[str] = None,
        resources: Optional[dict] = None,
        env_vars: Optional[dict[str, str]] = None,
        ports: Optional[list[dict]] = None,
        git_repo_url: Optional[str] = None,
    ) -> dict:
        """Create a new environment"""
        data = {"name": name, "template_id": template_id}
        if description:
            data["description"] = description
        if resources:
            data["resources"] = resources
        if env_vars:
            data["env_vars"] = env_vars
        if ports:
            data["ports"] = ports
        if git_repo_url:
            data["git_repo_url"] = git_repo_url
        return self._client.request("POST", "/environments", json=data)

    def update(self, id: str, **kwargs) -> dict:
        """Update an environment"""
        return self._client.request("PUT", f"/environments/{id}", json=kwargs)

    def delete(self, id: str) -> None:
        """Delete an environment"""
        self._client.request("DELETE", f"/environments/{id}")

    def start(self, id: str) -> dict:
        """Start an environment"""
        return self._client.request("POST", f"/environments/{id}/start")

    def stop(self, id: str) -> dict:
        """Stop an environment"""
        return self._client.request("POST", f"/environments/{id}/stop")

    def restart(self, id: str) -> dict:
        """Restart an environment"""
        return self._client.request("POST", f"/environments/{id}/restart")

    def get_ssh_config(self, id: str) -> dict:
        """Get SSH configuration"""
        return self._client.request("GET", f"/environments/{id}/ssh-config")

    def get_metrics(
        self,
        id: str,
        start_time: Optional[datetime] = None,
        end_time: Optional[datetime] = None,
    ) -> dict:
        """Get environment metrics"""
        params = {}
        if start_time:
            params["start_time"] = start_time.isoformat()
        if end_time:
            params["end_time"] = end_time.isoformat()
        return self._client.request("GET", f"/environments/{id}/metrics", params=params)


class AsyncEnvironmentsService:
    """Environments service for async client"""

    def __init__(self, client: "AsyncDevBoxClient"):
        self._client = client

    async def list(
        self,
        page: int = 1,
        page_size: int = 20,
        status: Optional[str] = None,
        template_id: Optional[str] = None,
    ) -> dict:
        """List environments"""
        params = {"page": page, "page_size": page_size}
        if status:
            params["status"] = status
        if template_id:
            params["template_id"] = template_id
        return await self._client.request("GET", "/environments", params=params)

    async def get(self, id: str) -> dict:
        """Get environment by ID"""
        return await self._client.request("GET", f"/environments/{id}")

    async def create(
        self,
        name: str,
        template_id: str,
        description: Optional[str] = None,
        resources: Optional[dict] = None,
        env_vars: Optional[dict[str, str]] = None,
        ports: Optional[list[dict]] = None,
        git_repo_url: Optional[str] = None,
    ) -> dict:
        """Create a new environment"""
        data = {"name": name, "template_id": template_id}
        if description:
            data["description"] = description
        if resources:
            data["resources"] = resources
        if env_vars:
            data["env_vars"] = env_vars
        if ports:
            data["ports"] = ports
        if git_repo_url:
            data["git_repo_url"] = git_repo_url
        return await self._client.request("POST", "/environments", json=data)

    async def update(self, id: str, **kwargs) -> dict:
        """Update an environment"""
        return await self._client.request("PUT", f"/environments/{id}", json=kwargs)

    async def delete(self, id: str) -> None:
        """Delete an environment"""
        await self._client.request("DELETE", f"/environments/{id}")

    async def start(self, id: str) -> dict:
        """Start an environment"""
        return await self._client.request("POST", f"/environments/{id}/start")

    async def stop(self, id: str) -> dict:
        """Stop an environment"""
        return await self._client.request("POST", f"/environments/{id}/stop")

    async def restart(self, id: str) -> dict:
        """Restart an environment"""
        return await self._client.request("POST", f"/environments/{id}/restart")

    async def get_ssh_config(self, id: str) -> dict:
        """Get SSH configuration"""
        return await self._client.request("GET", f"/environments/{id}/ssh-config")

    async def get_metrics(
        self,
        id: str,
        start_time: Optional[datetime] = None,
        end_time: Optional[datetime] = None,
    ) -> dict:
        """Get environment metrics"""
        params = {}
        if start_time:
            params["start_time"] = start_time.isoformat()
        if end_time:
            params["end_time"] = end_time.isoformat()
        return await self._client.request("GET", f"/environments/{id}/metrics", params=params)
